got: update to 0.111

artkiver opened the pull request:
#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 15.4.1 24E263 arm64
Command Line Tools 16.3.0.0.1.1742442376

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

aeiouaeiouaeiouaeiouaeiouaeiou commented:
Locally it builds fine, isn't it? I don't understand why CI can't get the package if it exists.

artkiver commented:
> Locally it builds fine, isn't it? I don't understand why CI can't get the package if it exists.

Yeah, builds OK locally. I am not sure why CI couldn't fetch it either?

There was something similar with the openssh PR I submitted, looked as it if failed trying to fetch ssh-copy-id? I dunno!

I was going to give another go at trying to set up mpbb locally and see if I could try to reproduce these errors locally, but the last couple of times I attempted to follow the mpbb instructions I seemed to get stuck and haven't gotten around to it yet.

But, maybe there's something else obvious I am missing? Or maybe the CI was just having some issues? (I think I saw another PR with failed builds around the same time as the two I submitted the other night)?

neverpanic commented:
CI fails because gameoftrees.org seems to reject connections from GitHub Actions:

```
  DEBUG: Fetching distfile failed: Failed to connect to gameoftrees.org port 443 after 30137 ms: Timeout was reached
  Error: Failed to fetch got: Failed to connect to gameoftrees.org port 443 after 30137 ms: Timeout was reached
```

artkiver commented:
OK, using inspiration from @stspdotname and thanks to @thomasadam for having already mirrored Game of Trees here: https://github.com/ThomasAdam/got-portable

I've modified the Portfile slightly and added:

PortGroup           github 1.0
github.setup        ThomasAdam got-portable 0.111
github.tarball_from tarball

`port -v fetch
--->  Fetching distfiles for got
--->  got-portable-0.111.tar.gz does not exist in /opt/local/var/macports/distfiles/got
--->  Attempting to fetch got-portable-0.111.tar.gz from http://mirror.fcix.net/macports/distfiles/got
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   153    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
--->  Attempting to fetch got-portable-0.111.tar.gz from https://distfiles.macports.org/got
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   126    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
--->  Attempting to fetch got-portable-0.111.tar.gz from https://codeload.github.com/ThomasAdam/got-portable/legacy.tar.gz/0.111?dummy=
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1180k    0 1180k    0     0  1773k      0 --:--:-- --:--:-- --:--:-- 1772k`

That all looks good.

This part not so much, but I swear I have run into something similar elsewhere in the past?

`port -v checksum
--->  Verifying checksums for got
--->  Checksumming got-portable-0.111.tar.gz
Error: Checksum (rmd160) mismatch for got-portable-0.111.tar.gz
Portfile checksum: got-portable-0.111.tar.gz rmd160 dc9ef32d736cb32cff7010baf35bb86e0aeb492e
Distfile checksum: got-portable-0.111.tar.gz rmd160 a9e56f9ada06b1d0fde8a65f9835c1ccc928bbda
Error: Checksum (sha256) mismatch for got-portable-0.111.tar.gz
Portfile checksum: got-portable-0.111.tar.gz sha256 d096f76e91a700dd0d22fbaf9641c2b94f8a6de16f09b0f4939c9b96a9d878ce
Distfile checksum: got-portable-0.111.tar.gz sha256 f8b103d7161e2e5ca54cca2953987b96e4757bd05a1c682230e4254fe1171eb7
Error: Checksum (size) mismatch for got-portable-0.111.tar.gz
Portfile checksum: got-portable-0.111.tar.gz size 1566105
Distfile checksum: got-portable-0.111.tar.gz size 1208841
The correct checksum line may be:
checksums           rmd160  a9e56f9ada06b1d0fde8a65f9835c1ccc928bbda \
                    sha256  f8b103d7161e2e5ca54cca2953987b96e4757bd05a1c682230e4254fe1171eb7 \
                    size    1208841
Error: Failed to checksum got: Unable to verify file checksums
Error: See /opt/local/var/macports/logs/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_devel_got/got/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there
is a bug.
Error: Processing of port got failed
`


Even more confusing, when I grab the tarball from https://github.com/ThomasAdam/got-portable/archive/refs/tags/0.111.tar.gz and check things manually I see:

`% openssl dgst -rmd160 got-portable-0.111.tar.gz
RIPEMD160(got-portable-0.111.tar.gz)= 32311c8279b51d7740b72c20acb1cd17cc521960
% openssl dgst -sha256 got-portable-0.111.tar.gz
SHA256(got-portable-0.111.tar.gz)= 494e34cc00444db0b1caa75e3fee6391d2401a06ebff8a74221721f42baa21a1
% ls -alF got-portable-0.111.tar.gz
-rw-r--r--@ 1 grey  staff  1208404 May  2 19:10 got-portable-0.111.tar.gz`

As contrasted with the got-portable-0.111.tar.gz I grabbed from https://gameoftrees.org/releases/portable/got-portable-0.111.tar.gz which is more like:

`
% openssl dgst -rmd160 got-portable-0.111.tar.gz
RIPEMD160(got-portable-0.111.tar.gz)= dc9ef32d736cb32cff7010baf35bb86e0aeb492e
% openssl dgst -sha256 got-portable-0.111.tar.gz
SHA256(got-portable-0.111.tar.gz)= d096f76e91a700dd0d22fbaf9641c2b94f8a6de16f09b0f4939c9b96a9d878ce
ls -alF got-portable-0.111.tar.gz
-rw-r--r--@ 1 grey  staff  1566105 Apr 25 05:58 got-portable-0.111.tar.gz
`

The "simple" fix is probably just to grab the suggestions from port -v checksum and amend the Portfile with those values, but it does make me wonder why GitHub seems to be serving up tarballs that have different sizes and hashes?

I will extract and check the tarballs manually, but I am also going to upload them here for others to check my work and see if anything weird is going on (I'm guessing it's entirely possible GitHub just adds some extra whatever that ends up changing the values, but since this isn't the first time I have seen this kind of thing, I don't want to not show my work just in case I am stumbling over something else. Also, I think it's about time I grab some lunch so will be AFK and not able to address this further for a little bit longer).
[gottarballsforhashinganddiffing.tar.gz](https://github.com/user-attachments/files/20017480/gottarballsforhashinganddiffing.tar.gz)

artkiver commented:
[Portfile.got-portable0.111_WIP_GitHub.txt](https://github.com/user-attachments/files/20017553/Portfile.got-portable0.111_WIP_GitHub.txt)
(Here's the Work in Progress Portfile just for others to have that as a reference too; though it has the hash values generated from the gameoftrees.org downloaded file rather than the GitHub tarballs so it won't function as is just yet)

artkiver commented:
Looks as if the structure of the GitHub tarball from ThomasAdam's repo is slightly different than the upstream -portable tarball.

So there's no configure script by default, unless sh autogen.sh has been instantiated first.

Let me see if I can figure out how to rectify that.

artkiver commented:
Wow, that last commit finally seemed to make GitHub Actions CI happy?

Though now the following is probably redundant and can probably be removed:

`    # these dependencies are necessary to re-detect the headers
    depends_build-append \
                        port:autoconf \
                        port:automake
`

artkiver commented:
As discussed on the FediVerse with Stefan and Thomas, my methodology is probably not a great way to address the lack of mirrors usable by GitHub Actions long term; but hopefully will suffice for the time being until I can sort out some saner strategy.

ThomasAdam commented:
Hi,

I'm not sure what's happening here, but I really wouldn't suggest this is the best approach.  If anything, it's the exact opposite.

Because I mirror the code on Github -- it's just that; a mirror.  This will of course include things like tags, etc., as part of that.

But the *dist tarball* created from that is slightly different from the tagged commit -- because in running `make dist`. that doesn't include certain files.  This is important, since a tarball created from a tag, vs., a proper release will contain different thing; the execution of `./configure` will vary, for instance.

So I hope I've read this correctly -- but if you're suggesting that taking a tarball from a known tag from Github is considered a "release", it really isn't.

The only canonical place for release right now is on gameoftrees.org

If this is a problem, I can look at mirroring that, but it would be just that; a mirror of released tarballs.

But if your approach for now is to think you can use GitHub for -portable releases, please stop.

Seriously.

It's not only misleading, it's just incorrect.

artkiver commented:
> Hi,
> 
> I'm not sure what's happening here, but I really wouldn't suggest this is the best approach. If anything, it's the exact opposite.
> 
> Because I mirror the code on Github -- it's just that; a mirror. This will of course include things like tags, etc., as part of that.
> 
> But the _dist tarball_ created from that is slightly different from the tagged commit -- because in running `make dist`. that doesn't include certain files. This is important, since a tarball created from a tag, vs., a proper release will contain different thing; the execution of `./configure` will vary, for instance.
> 
> So I hope I've read this correctly -- but if you're suggesting that taking a tarball from a known tag from Github is considered a "release", it really isn't.
> 
> The only canonical place for release right now is on gameoftrees.org
> 
> If this is a problem, I can look at mirroring that, but it would be just that; a mirror of released tarballs.
> 
> But if your approach for now is to think you can use GitHub for -portable releases, please stop.
> 
> Seriously.
> 
> It's not only misleading, it's just incorrect.

That's a completely reasonable response.

I'm going to close this for the time being and revisit it with better mirror alternatives sometime in the future.

Just because I can do some strange things to make MacPorts' use of GitHub Actions "happy" doesn't mean that's a good way to proceed.

Thanks for chiming in with your perspective!

ryandesign commented:
> So I hope I've read this correctly -- but if you're suggesting that taking a tarball from a known tag from Github is considered a "release", it really isn't.
> 
> The only canonical place for release right now is on gameoftrees.org
> 
> If this is a problem, I can look at mirroring that, but it would be just that; a mirror of released tarballs.
> 
> But if your approach for now is to think you can use GitHub for -portable releases, please stop.
> 
> Seriously.
> 
> It's not only misleading, it's just incorrect.

@ThomasAdam Absolutely agree. MacPorts should use release tarballs when developers offer them.

You can [create a GitHub release](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases) for each of your tags. In addition to the two source code download links that GitHub presents automatically which are not release downloads and which you unfortunately cannot turn off, you can attach your manually-created release tarball to each release.

ThomasAdam commented:
@ryandesign 

Sure -- I'm well aware of how to handle releases on GitHub, including assets.

But I won't do that for gameoftrees.

It's a *code* mirror only.

What I can do, is mirror the -portable release elsewhere.

However, before I do that, I will need to look at signing those releases so that mirrors are "safer" with that information.  This is something we're not doing at the moment.

stspdotname commented:
Hi, upstream gameoftrees maintainer here. I have provided my opinion on mastodon before, but since this discussion seems to be going in circles, I will do it again:

As far as I understand, the situation is as follows:

1) macports uses github for CI purposes and the CI build wants to download a release tarball when it runs.

2) Our release tarballs are hosted at a volunteer-driven non-profit ISP in Berlin, who has recently started blocking Microsoft IPs (which includes github) because requests from Microsoft's network were causing unreasonable load on the non-proft ISP's web servers.

These two circumstances are unfortunately in conflict with each other.

Given that Macports has chosen to rely on Github for CI and is now getting caught up in the fallout of bad practices and misbehaviour of Github's owner, I do not believe it is fair to expect that the gameoftrees project (myself and Thomas Adam included) will do anything to fix this problem for you. We are volunteers and our time is limited.

A reasonable workaround would be that Macport maintainers create their own mirror of the release artifacts published by us. As pointed out above, anyone can upload artifacts to Github somewhere. Uploading the artifacts to github would side-step the block and should always keep working since Github is unlikely to block itself.

What follows is a quote of what I already said earlier on mastodon:

For stable checksums you need actual copies of tarballs downloaded from gameoftrees.org/releases. Github won't provide those tarballs automatically. You will have to upload them to Github yourself somehow. E.g. upload them into a Git repo, or create fake game of trees releases in a fork of Thomas Adam's github repo where you can then upload our tarballs as what Github calls "binaries" associated with the release.  Or maybe tarballs could be uploaded to a github-pages web site and appear as static assets for download there.  There are probably even more ways to achieve this. Just pick some approach that works for you.

aeiouaeiouaeiouaeiouaeiouaeiou commented:
I think this pull request is closed too soon.
I have an idea to add [FreeBSD mirrors](https://pkg.freebsd.org/ports-distfiles/got-portable-0.111.tar.gz) to the portfile (they are a bit behind upstream, but not afraid of LLM bots).

artkiver commented:
> I think this pull request is closed too soon. I have an idea to add [FreeBSD mirrors](https://pkg.freebsd.org/ports-distfiles/got-portable-0.111.tar.gz) to the portfile (they are a bit behind upstream, but not afraid of LLM bots).

The FreeBSD mirrors I hadn't considered, but it seems to be functioning! Thanks for this Pull Request @aeiouaeiouaeiouaeiouaeiouaeiou https://github.com/macports/macports-ports/pull/28464 tested locally OK after running port clean --dist
